### ISSUE TITLE
Where can i find the reference for nv greamstream protocol?

### ISSUE THREAD

**Issue by hitner** [moonlight-stream/moonlight-common-c#67]
As the title means,  does the protocol is free to use and open to public? I cannot find more infomation on NV web.

**Comment by cgutman** [comment]
The Nvidia GameStream protocol is proprietary. There is no reference documentation.

Our implementation is based on reverse engineering. The best public reference is this repository, Moonlight clients like https://github.com/moonlight-stream/moonlight-embedded, and Moonlight-compatible hosting software like https://github.com/loki-47-6F-64/sunshine

**Comment by ReenigneArcher** [comment]
@cgutman sunshine is poorly documented and the code is a mess. The best documentation anyone can provide is some hand written notes that no one can understand. On top of that Loki is no longer involved in the project, or just ignoring the entire project. The project is hanging on by a thread.

In my opinion, it would be very beneficial for moonlight to document the protocol. This article (https://www.writethedocs.org/guide/writing/beginners-guide-to-docs/) describes many reasons why it's a good idea to document your project.

I started a [thread](https://discord.com/channels/352065098472488960/842462527257968670/961689433848045621) on the discord, but never received a response. Seems you are the all knowing one on the protocol, so documenting it should be an effort led by you although I'm sure many are willing to assist. For example, I'd be willing to setup the sphinx documentation framework and integrations with github.

I understand that having the code itself somewhat serves as documentation; however it's pretty difficult to read through. One major issue with it is it expects the person trying to understand the protocols understands C/C++ or whatever language the code was written in. For example, I want to investigate writing a python based host for moonlight. Currently my only options are to reverse engineer moonlight, or reverse engineer sunshine (which in itself reverse engineered moonlight).

Documenting the protocol has been asked for a lot, ideally it would be in human readable terms. For example the documentation of pairing would describe:
 - what types of request to sent to and from the server
 - what port and endpoint
 - what are the expected headers
 - how to authenticate
 - the order of the requests
 - etc.
 
The protocol documentation should also describe how to handle things on the both the client and host side. For example with a discovery request, a moonlight client will require different requests as compared to the host.

Other related issues:
https://github.com/moonlight-stream/moonlight-common-c/issues/69
https://github.com/moonlight-stream/moonlight-common-c/issues/41
https://github.com/moonlight-stream/moonlight-common-c/issues/29
https://github.com/moonlight-stream/moonlight-chrome/issues/510#issuecomment-908237356

**Comment by cgutman** [comment]
Yes, I can try to fill in details if you can create a doc template or something.

There are 6 main protocols involved in the current GameStream protocol version:
- HTTP (TCP 47989)
- HTTPS (TCP 47984)
- RTSP (TCP 48010)
- Control over [ENet](http://enet.bespin.org/) (UDP 47999)
- Video over RTP (UDP 47998)
- Audio over RTP (UDP 48000)

For the first 3, doc templates designed for HTTP-based protocols should work.

The latter 3 will probably need templates designed for binary protocols (something [like this](https://en.wikipedia.org/wiki/Transmission_Control_Protocol#TCP_segment_structure))

**Comment by ReenigneArcher** [comment]
Here's the initial framework for the docs. Definitely no where near ready.

https://moonlight-protocol.readthedocs.io/en/nightly/index.html

https://github.com/SunshineStream/moonlight-protocol/tree/nightly
